### PR TITLE
[WIP] Avoid using thread to execute methods

### PR DESCRIPTION
**Pull request by sergiyv-bitquill**
### Summary

Avoid using threads to execute JDBC methods.

### Description

Monitoring threads can abort a current connection that's waiting for a server response.

### Additional Reviewers

@matthewh-BQ 
@karenc-bq 
@ColinKYuen 
@brunos-bq

### CODE REVIEW

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java [hunk 1]
```diff
@@ -136,48 +132,43 @@ public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> ex
...
 
-      result = executeFuncFuture.get();
-    } catch (ExecutionException exception) {
-      final Throwable throwable = exception.getCause();
-      if (throwable instanceof Error) {
-        throw (Error) throwable;
+    } catch (SQLException sqlEx) {
+      throw sqlEx;
+    } catch (CJException cjEx) {
+      throw cjEx;
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java:160]:
```suggestion
} catch (SQLException | CJException exception) {
      throw exception;
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java [hunk 2]
```diff
@@ -136,48 +132,43 @@ public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> ex
...
+      } else {
+        throw ex;
       }
-      throw (Exception) throwable;
     } finally {
-      this.monitorService.stopMonitoring(this.monitorContext);
-      if (executor != null) {
-        executor.shutdownNow();
+      if (monitorContext != null) {
+        monitorContext.getLock().lock();
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java:170]: Why are we locking this twice? Should this be `unlock()` instead?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java [hunk 3]
```diff
@@ -26,9 +26,12 @@
 
 package com.mysql.cj.jdbc.ha.ca.plugins;
 
+import com.mysql.cj.jdbc.JdbcConnection;
 import com.mysql.cj.log.Log;
 
+import java.sql.SQLException;
 import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java:34]: Unused import

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java [hunk 4]
```diff
@@ -40,11 +42,6 @@
 import java.util.HashSet;
 import java.util.Set;
 import java.util.concurrent.Callable;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
-import java.util.concurrent.TimeUnit;
 
 public class NodeMonitoringConnectionPlugin implements IConnectionPlugin {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java:47]: `static final int CHECK_INTERVAL_MILLIS = 1000;` no longer being used

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java [hunk 5]
```diff
@@ -101,7 +101,10 @@ public synchronized void stopMonitoring(MonitorConnectionContext context) {
       log.logWarn(NullArgumentException.constructNullArgumentMessage("context"));
       return;
     }
-    this.contexts.remove(context);
+    synchronized (context) {
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/Monitor.java:104]: Method signature is also `synchronized`

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java [hunk 6]
```diff
@@ -46,6 +46,7 @@
   private long invalidNodeStartTime;
   private int failureCount;
   private boolean nodeUnhealthy;
+  private boolean activeContext = true;
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java:49]: I think we should use a volatile here

Comment (karenc-bq) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java:49]: Actually volatile wouldn't really contribute

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java [hunk 7]
```diff
@@ -136,48 +132,44 @@ public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> ex
...
 
-      result = executeFuncFuture.get();
-    } catch (ExecutionException exception) {
-      final Throwable throwable = exception.getCause();
-      if (throwable instanceof Error) {
-        throw (Error) throwable;
+    } catch (SQLException | CJException exception) {
+      throw exception;
+    } catch (Exception ex) {
+      if (monitorContext != null && monitorContext.isNodeUnhealthy()) {
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java:158]: We don't need the `catch (Exception ex)` clause if we add the checks to finally.

**Review by ColinKYuen — APPROVED** [review]

**Pull request merged**